File: engines/lancedb-tantivy/do_query.py
#!/usr/bin/env python3
"""
优化版 LanceDB Tantivy FTS 查询实现
主要优化点:
1. 使用 select([]) 只返回必要字段，避免传输 text 内容
2. 避免使用 to_list()，直接操作 Arrow 数据
3. 支持 Tantivy 风格的查询语法转换
"""
import warnings
import sys
import os
import re
import lancedb
from pathlib import Path

# 通过环境变量控制调试输出
DEBUG = os.environ.get('DEBUG', '0') == '1'

def debug_print(*args, **kwargs):
    """仅在DEBUG模式下打印调试信息"""
    if DEBUG:
        print(*args, **kwargs)


def parse_query(query_str):
    """
    解析 Tantivy 风格的查询字符串，返回 (query_type, terms)
    
    支持的格式:
    - term: "the" -> ("term", "the")
    - union: "word1 word2" -> ("union", "word1 word2")  
    - intersection: "+word1 +word2" -> ("intersection", "word1 word2")
    - phrase: "\"word1 word2\"" -> ("phrase", "word1 word2")
    """
    query_str = query_str.strip()
    
    # 检查是否是 phrase 查询 (被双引号包围)
    if query_str.startswith('"') and query_str.endswith('"'):
        # 移除外层引号
        inner = query_str[1:-1]
        return ("phrase", inner)
    
    # 检查是否是 intersection 查询 (所有词都有 + 前缀)
    # 模式: +word1 +word2 +word3
    if query_str.startswith('+'):
        # 提取所有 +term
        terms = re.findall(r'\+(\S+)', query_str)
        if terms:
            # 检查整个查询是否只由 +term 组成
            reconstructed = ' '.join(['+' + t for t in terms])
            if reconstructed == query_str:
                return ("intersection", ' '.join(terms))
    
    # 检查是否是单个词 (term 查询)
    words = query_str.split()
    if len(words) == 1:
        return ("term", query_str)
    
    # 默认是 union 查询 (多个词，用空格分隔)
    return ("union", query_str)


def execute_fts_query(table, query_str, limit):
    """
    根据查询类型执行 FTS 查询
    
    根据 LanceDB 文档 (https://docs.lancedb.com/search/full-text-search):
    - phrase_query(True): 精确短语匹配
    - phrase_query(False): 词汇匹配 (默认 OR)
    """
    query_type, terms = parse_query(query_str)
    
    debug_print(f"DEBUG: Parsed query: type={query_type}, terms={terms}", file=sys.stderr, flush=True)
    
    if query_type == "phrase":
        # Phrase 查询: 使用 phrase_query(True)
        return table.search(terms, query_type="fts").phrase_query(True).select(["id"]).limit(limit)
    
    elif query_type == "intersection":
        # Intersection (AND) 查询
        # LanceDB 默认是 OR，这里暂时用 OR 查询
        return table.search(terms, query_type="fts").phrase_query(False).select(["id"]).limit(limit)
    
    elif query_type == "union":
        # Union (OR) 查询: 默认行为
        return table.search(terms, query_type="fts").phrase_query(False).select(["id"]).limit(limit)
    
    else:
        # Term 查询: 单个词
        return table.search(terms, query_type="fts").select(["id"]).limit(limit)


def main():
    if len(sys.argv) < 2:
        print("Usage: do_query.py <index_dir>", file=sys.stderr)
        sys.exit(1)
    
    index_dir = Path(sys.argv[1])
    
    # 连接到LanceDB数据库
    debug_print(f"DEBUG: Connecting to database at {index_dir}", file=sys.stderr, flush=True)
    db = lancedb.connect(index_dir)
    table = db.open_table("wiki_articles")
    debug_print("DEBUG: Database connection successful", file=sys.stderr, flush=True)
    
    for line in sys.stdin:
        line = line.strip()
        if not line:
            continue
        
        fields = line.split("\t")
        if len(fields) != 2:
            print("UNSUPPORTED", flush=True)
            continue
        
        command = fields[0]
        query_str = fields[1]
        
        debug_print(f"DEBUG: Processing command={command}, query={query_str}", file=sys.stderr, flush=True)
        
        try:
            if command == "COUNT":
                result = execute_fts_query(table, query_str, 1000000).to_arrow()
                print(result.num_rows, flush=True)
            
            elif command == "TOP_10":
                result = execute_fts_query(table, query_str, 10).to_arrow()
                print(result.num_rows if result.num_rows > 0 else 1, flush=True)
            
            elif command == "TOP_100":
                result = execute_fts_query(table, query_str, 100).to_arrow()
                print(result.num_rows if result.num_rows > 0 else 1, flush=True)
            
            elif command == "TOP_1000":
                result = execute_fts_query(table, query_str, 1000).to_arrow()
                print(result.num_rows if result.num_rows > 0 else 1, flush=True)
            
            elif command == "TOP_1_COUNT":
                result = execute_fts_query(table, query_str, 1).to_arrow()
                count_result = execute_fts_query(table, query_str, 100000).to_arrow()
                print(count_result.num_rows, flush=True)
            
            elif command == "TOP_5_COUNT":
                result = execute_fts_query(table, query_str, 5).to_arrow()
                count_result = execute_fts_query(table, query_str, 100000).to_arrow()
                print(count_result.num_rows, flush=True)
            
            elif command == "TOP_10_COUNT":
                result = execute_fts_query(table, query_str, 10).to_arrow()
                count_result = execute_fts_query(table, query_str, 100000).to_arrow()
                print(count_result.num_rows, flush=True)
            
            elif command == "TOP_100_COUNT":
                result = execute_fts_query(table, query_str, 100).to_arrow()
                count_result = execute_fts_query(table, query_str, 100000).to_arrow()
                print(count_result.num_rows, flush=True)
            
            elif command == "TOP_1000_COUNT":
                result = execute_fts_query(table, query_str, 1000).to_arrow()
                count_result = execute_fts_query(table, query_str, 100000).to_arrow()
                print(count_result.num_rows, flush=True)
            
            else:
                print("UNSUPPORTED", flush=True)
        
        except Exception as e:
            debug_print(f"ERROR: {e}", file=sys.stderr, flush=True)
            print("UNSUPPORTED", flush=True)

if __name__ == "__main__":
    main()

File: engines/lancedb-rust-fts/src/bin/do_query.rs
use lancedb::connect;
use lancedb::query::{ExecutableQuery, QueryBase, Select};
use lancedb::index::scalar::{
    FtsQuery, FullTextSearchQuery, MatchQuery,
    Operator, PhraseQuery};
use lancedb::Table;
use std::env;
use std::io::BufRead;
use std::path::Path;
use once_cell::sync::Lazy;
use regex::Regex;

// 预编译正则表达式
static INTERSECTION_RE: Lazy<Regex> = Lazy::new(|| Regex::new(r"\+(\S+)").unwrap());

// 预分配的列选择
static ID_COLUMNS: Lazy<Select> = Lazy::new(|| Select::Columns(vec!["id".to_string()]));

/// 查询类型枚举
#[derive(Debug)]
enum QueryType {
    Term(String),           // 单个词
    Union(String),          // OR 查询 (word1 word2)
    Intersection(String),   // AND 查询 (+word1 +word2)
    Phrase(String),         // 短语查询 ("word1 word2")
}

/// 解析 Tantivy 风格的查询字符串
#[inline]
fn parse_query(query_str: &str) -> QueryType {
    let query_str = query_str.trim();
    
    // 检查是否是 phrase 查询 (被双引号包围)
    if query_str.starts_with('"') && query_str.ends_with('"') && query_str.len() > 2 {
        let inner = &query_str[1..query_str.len()-1];
        return QueryType::Phrase(inner.to_string());
    }
    
    // 检查是否是 intersection 查询 (所有词都有 + 前缀)
    if query_str.starts_with('+') {
        let terms: Vec<String> = INTERSECTION_RE.captures_iter(query_str)
            .filter_map(|cap| cap.get(1).map(|m| m.as_str().to_string()))
            .collect();
        
        if !terms.is_empty() {
            // 检查整个查询是否只由 +term 组成
            let reconstructed = terms.iter()
                .map(|t| format!("+{}", t))
                .collect::<Vec<_>>()
                .join(" ");
            
            if reconstructed == query_str {
                return QueryType::Intersection(terms.join(" "));
            }
        }
    }
    
    // 检查是否是单个词 (term 查询)
    if !query_str.contains(' ') {
        return QueryType::Term(query_str.to_string());
    }
    
    // 默认是 union 查询 (多个词，用空格分隔)
    QueryType::Union(query_str.to_string())
}

/// 根据查询类型构建 FTS 查询
#[inline]
fn build_fts_query(query_str: &str) -> FullTextSearchQuery {
    let query_type = parse_query(query_str);
    
    let fts_query = match query_type {
        QueryType::Phrase(terms) => {
            FtsQuery::Phrase(PhraseQuery::new(terms))
        }
        QueryType::Intersection(terms) => {
            FtsQuery::Match(MatchQuery::new(terms).with_operator(Operator::And))
        }
        QueryType::Union(terms) => {
            FtsQuery::Match(MatchQuery::new(terms).with_operator(Operator::Or))
        }
        QueryType::Term(term) => {
            FtsQuery::Match(MatchQuery::new(term))
        }
    };
    
    FullTextSearchQuery::new_query(fts_query)
}

/// 执行查询并返回结果数量
#[inline]
async fn execute_count(table: &Table, query_str: &str, limit: usize) -> Result<usize, ()> {
    let fts_query = build_fts_query(query_str);
    let query_result = table
        .query()
        .full_text_search(fts_query)
        .select(ID_COLUMNS.clone())
        // .select(Select::columns(&["_score"]))  // 只选择 FTS 分数列
        // .with_row_id()
        // .fast_search()
        .limit(limit)
        .execute()
        .await;
    
    match query_result {
        Ok(results) => {
            use futures::StreamExt;
            let mut count = 0usize;
            futures::pin_mut!(results);
            while let Some(batch_result) = results.next().await {
                if let Ok(batch) = batch_result {
                    count += batch.num_rows();
                }
            }
            Ok(count)
        }
        Err(_) => Err(())
    }
}

/// 执行 TOP N 查询
#[inline]
async fn execute_top_n(table: &Table, query_str: &str, limit: usize) -> Result<(), ()> {
    let fts_query = build_fts_query(query_str);
    let query_result = table
        .query()
        .full_text_search(fts_query)
        .select(ID_COLUMNS.clone())
        // .select(Select::columns(&["_score"]))  // 只选择 FTS 分数列
        // .with_row_id()
        // .fast_search()
        .limit(limit)
        .execute()
        .await;
    
    match query_result {
        Ok(results) => {
            use futures::StreamExt;
            futures::pin_mut!(results);
            // 消费流但不需要收集
            while let Some(_) = results.next().await {}
            Ok(())
        }
        Err(_) => Err(())
    }
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let args: Vec<String> = env::args().collect();
    let index_dir = Path::new(&args[1]);
    
    // 连接到 LanceDB
    let db = connect(index_dir.to_str().unwrap()).execute().await?;
    let table = db.open_table("wiki_articles").execute().await?;
    
    let stdin = std::io::stdin();
    for line in stdin.lock().lines() {
        let line = line?;
        let fields: Vec<&str> = line.split('\t').collect();
        
        if fields.len() != 2 {
            println!("UNSUPPORTED");
            continue;
        }
        
        let command = fields[0];
        let query_str = fields[1];
        
        match command {
            "COUNT" => {
                match execute_count(&table, query_str, 1000000).await {
                    Ok(count) => println!("{}", count),
                    Err(_) => println!("UNSUPPORTED"),
                }
            }
            "TOP_10" => {
                match execute_top_n(&table, query_str, 10).await {
                    Ok(_) => println!("1"),
                    Err(_) => println!("UNSUPPORTED"),
                }
            }
            "TOP_100" => {
                match execute_top_n(&table, query_str, 100).await {
                    Ok(_) => println!("1"),
                    Err(_) => println!("UNSUPPORTED"),
                }
            }
            "TOP_1000" => {
                match execute_top_n(&table, query_str, 1000).await {
                    Ok(_) => println!("1"),
                    Err(_) => println!("UNSUPPORTED"),
                }
            }
            "TOP_1_COUNT" | "TOP_5_COUNT" | "TOP_10_COUNT" | "TOP_100_COUNT" | "TOP_1000_COUNT" => {
                let limit = match command {
                    "TOP_1_COUNT" => 1,
                    "TOP_5_COUNT" => 5,
                    "TOP_10_COUNT" => 10,
                    "TOP_100_COUNT" => 100,
                    "TOP_1000_COUNT" => 1000,
                    _ => 10,
                };
                
                // 获取 top N (消费结果)
                if execute_top_n(&table, query_str, limit).await.is_err() {
                    println!("UNSUPPORTED");
                    continue;
                }
                
                // 获取总数
                match execute_count(&table, query_str, 100000).await {
                    Ok(count) => println!("{}", count),
                    Err(_) => println!("UNSUPPORTED"),
                }
            }
            _ => {
                println!("UNSUPPORTED");
            }
        }
    }
    
    Ok(())
}